### ISSUE TITLE
CoinSort.hpp compilation issue

### ISSUE THREAD

**Issue by svigerske** [coin-or/CoinUtils#64]
Issue created by migration from Trac.

Original creator: girardea

Original creation time: 2011-05-24 16:39:00

Version: 

I just downloaded the latest tarball of Clp (1.13.3), not sure which version of CoinUtils is included in there. When trying to compile a file that includes CoinSort.hpp (indirectly through OsiClpSolverInterface.hpp) with GCC4.5.2 or GCC4.2, I get the following errors:

```
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp: In member function ‘bool CoinFirstLess_2<S, T>::operator()(const CoinPair<S, T>&, const CoinPair<S, T>&) const’:
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:45:26: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:45:15: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp: In member function ‘bool CoinFirstAbsLess_2<S, T>::operator()(const CoinPair<S, T>&, const CoinPair<S, T>&) const’:
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:68:56: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:68:67: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:68:24: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:69:56: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:69:67: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:69:24: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp: In member function ‘bool CoinFirstAbsGreater_2<S, T>::operator()(CoinPair<S, T>, CoinPair<S, T>) const’:
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:82:56: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:82:67: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:82:24: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:83:56: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:83:67: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:83:24: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp: In member function ‘bool CoinFirstLess_3<S, T, U>::operator()(const CoinTriple<S, T, U>&, const CoinTriple<S, T, U>&) const’:
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:376:26: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:376:15: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp: In member function ‘bool CoinFirstAbsLess_3<S, T, U>::operator()(const CoinTriple<S, T, U>&, const CoinTriple<S, T, U>&) const’:
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:399:56: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:399:67: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:399:24: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:400:56: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:400:67: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:400:24: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp: In member function ‘bool CoinFirstAbsGreater_3<S, T, U>::operator()(const CoinTriple<S, T, U>&, const CoinTriple<S, T, U>&) const’:
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:414:56: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:414:67: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:414:24: error: parse error in template argument list
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:415:56: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:415:67: error: ‘.’ cannot appear in a constant-expression
/home/pierre/devel/Clp-1.13.3/build/include/coin/CoinSort.hpp:415:24: error: parse error in template argument list
```

It's been a while since I haven't used Clp, but I do not recall having that type of errors. It used to compile straight away. Is this a known bug? Should I consider downgrading to an earlier version in the meantime?

Thanks.

**Comment by svigerske** [comment]
Comment by @LouHafer created at 2011-05-24 16:43:02

I pulled down Clp-1.13.3 from svn and did a build with Sun Studio 12.1 and GCC 4.4.1 and both builds were clean. I'll post to the CoinUtils mailing list and see if anyone can test with the GCC versions you cite. Another suggestion (low probablility): try an uninstall of CoinUtils and rebuild, on the off chance that somehow an old copy of CoinSort.hpp is floating around.  -- Lou --

**Comment by svigerske** [comment]
Comment by girardea created at 2011-05-25 04:35:49

I think I found the mistake, which is quite entirely mine. It may help others though. It seems to be a conflict with blitz++. In some of my headers, I had put a "using namespace blitz;" before including the headers from Osi. For some reason it put the compilation error I mentioned before. This is generally a bad idea to use "using namespace" in a header, so I deleted it and the compilation went fine.

Thanks for your time and your replying so quickly!
